Update and expand theory sections (#3114)
ROCm/rocm-systems: projects/hip/docs/how-to/performance_guidelines.rst

   Repeat until performance goals are met.

Profiling and analysis tools
=============================

> ROCm offers a comprehensive suite of profiling and analysis tools that help developers understand and optimize GPU performance. These tools are essential for identifying bottlenecks and evaluating the effectiveness of performance optimizations.

Reviewer: ```suggestion
ROCm provides a comprehensive suite of profiling and analysis tools that help developers understand and optimize GPU performance. These tools are essential for identifying bottlenecks and evaluating the effectiveness of performance optimizations.
```

ROCm Profiler (rocprofv3)
-------------------------

> The ROCm Profiler (``rocprofv3``) provides command-line-driven profiling for

Reviewer: The first occurrence of the tool name in the body should be linked to the docs for the tool. 
I'll only call it out it once but applies below, too.

Author: Fixed.

detailed performance analysis. It collects metrics on kernel execution time,

> memory bandwidth, wavefront occupancy, VALU utilization, and instruction-level

Reviewer: VALU might be a good word to add to the Glossary. 
By the way, we should discuss how we will be maintaining the Glossary going forward. How do we decide what is glossary worthy?

Author: Added VALU and its siblings to the glossary.

counters.

``rocprofv3`` integrates with the rocProfiler-SDK and rocTracer frameworks to
collect hardware traces and API-level timing data. The collected data can be
exported in JSON and CSV formats for further analysis or visualization.

Key capabilities:

* Kernel execution profiling
* Memory bandwidth analysis
* Wavefront occupancy metrics
* Compute unit utilization
* Instruction-level performance counters
* API call tracing
* Hardware event collection

> For detailed usage instructions, see the :doc:`rocprofv3 documentation

Reviewer: Perhaps this line isn't needed if you provide the hyperlink above. 

If you keep it, refer to the actual component name and not rocprofv3

Author: rocprofv3 is the component name: https://rocm.docs.amd.com/projects/rocprofiler-sdk/en/latest/how-to/using-rocprofv3.html

<rocprofiler-sdk:how-to/using-rocprofv3>`.

ROCprof Compute Viewer
----------------------

ROCprof Compute Viewer provides a GUI-based environment for analyzing GPU
kernel performance data. It delivers detailed kernel‑level insights, including counter correlation and hierarchical performance breakdowns, helping developers interpret execution patterns and identify optimization opportunities.

Key features:

* Kernel performance counter analysis
* Counter correlation and visualization
* Hierarchical kernel performance breakdown
* Interactive performance data exploration
* Kernel-level optimization insights

For more information, see the
`ROCprof Compute Viewer documentation <https://rocm.docs.amd.com/projects/rocprof_compute_viewer/en/latest/>`__.

Perfetto trace viewer
---------------------

> 

Reviewer: This new section of the docs mentions that "ROCm provides"

<img width="1317" height="210" alt="Image" src="https://github.com/user-attachments/assets/21a8722d-d732-4498-8755-6ad66f1264c8" />

However, we're not providing this tool. Perhaps we should be clearer that this is a third-party tool and not from AMD or part of ROCm. 
Curious why we're calling out this third-party tool and not others.

Author: AFAIK, Perfetto is our go-to tool for timeline visualization, we don't provide our own solution. It is frequently mentioned as such in the profiler documentation.

Author: I merged Perfetto with rocprofv3, it is now just a subsection of the latter. And it is clearly mentioned that Perfetto is a third-party application.

For application-level tracing, including API calls, library calls, and
host-device synchronization, use the Perfetto trace viewer at

> https://ui.perfetto.dev. The ``rocprofv3`` tool can generate traces in

Reviewer: The highlight should be linked, and the redline should be removed.

<img width="1234" height="215" alt="Image" src="https://github.com/user-attachments/assets/f7130594-0414-4603-a04c-f97043041a26" />

Author: Fixed.

Perfetto-compatible format, allowing you to visualize the complete timeline of
application execution.

Perfetto provides a web‑based interface for analyzing traces, visualizing the temporal relationships among host operations, kernel launches, memory transfers, and synchronization events. This holistic perspective helps identify performance issues caused from API overhead, inefficient synchronization, or insufficient overlap between computation and data movement.

Key capabilities:

* Timeline visualization of application traces
* HIP API call tracking and timing
* ROCm library call analysis
* Host-device synchronization visualization
* Memory transfer tracking
* Interactive trace exploration and filtering

AMD System Management Interface
--------------------------------

> The AMD System Management Interface (``amd-smi``) is a command-line utility for

Reviewer: ```suggestion
The AMD System Management Interface (AMD SMI) is a command-line utility for
```

querying, monitoring, and managing AMD GPUs. It provides system administrators
and developers with detailed, real-time information about GPU hardware,
utilization, and power metrics.

``amd-smi`` reports the following categories of information:

> * GPU identity information such as the card name, device ID, and PCI bus

Reviewer: ```suggestion
* GPU identity information, such as the card name, device ID, and PCI bus
```

  location
* Live utilization metrics, including GPU activity, memory usage, clock speeds,
  and active processes
* Power and thermal readings, such as temperature, fan speed, voltage, and
  power draw
* Performance states and limits, such as available frequency levels, clock
  throttling, and voltage controls

> These metrics are retrieved through the ROCm SMI C API, which exposes a stable,

Reviewer: Is it:
ROCm SMI C API

or

AMD SMI C API

Reviewer 2: Fixed.

scriptable interface for system and performance monitoring tools.

``amd-smi`` also supports management operations, including:

* Setting power caps and performance profiles
* Adjusting clock frequencies
* Performing GPU resets and controlling persistence mode
* Reporting or controlling ECC status on supported data center GPUs

Output can be formatted as human-readable text or JSON (``--json``), commonly
used for integration into automated monitoring pipelines.

Basic usage examples:

.. code-block:: shell

   # Display GPU information
   amd-smi static

   # Monitor GPU utilization and metrics
   amd-smi metric --usage

   # Show detailed information in JSON format
   amd-smi static --json

For complete documentation, see the :doc:`AMD SMI documentation <amdsmi:index>`.

Typical workflow
----------------

For comprehensive ROCm performance analysis:

> 

Reviewer: I would use the component name here (highlight):

<img width="1262" height="486" alt="Image" src="https://github.com/user-attachments/assets/641ac868-5e1d-4f2b-9630-2cf5f90d1854" />

Author: Fixed.

   Repeat until performance goals are met.

Profiling and analysis tools
=============================

ROCm provides comprehensive profiling and analysis tools for understanding and
optimizing GPU performance. These tools are essential for diagnosing

> bottlenecks and measuring the impact of optimizations.

Reviewer 3: ```suggestion
ROCm offers a comprehensive suite of profiling and analysis tools that help developers understand and optimize GPU performance. These tools are essential for identifying bottlenecks and evaluating the effectiveness of performance optimizations.
```

ROCm Profiler (rocprofv3)
-------------------------

The ROCm Profiler (``rocprofv3``) provides command-line-driven profiling for
detailed performance analysis. It collects metrics on kernel execution time,
memory bandwidth, wavefront occupancy, VALU utilization, and instruction-level
counters.

``rocprofv3`` integrates with the rocProfiler-SDK and rocTracer frameworks to
collect hardware traces and API-level timing data. The collected data can be
exported in JSON and CSV formats for further analysis or visualization.

Key capabilities:

* Kernel execution profiling
* Memory bandwidth analysis
* Wavefront occupancy metrics
* Compute unit utilization
* Instruction-level performance counters
* API call tracing
* Hardware event collection

For detailed usage instructions, see the :doc:`rocprofv3 documentation
<rocprofiler-sdk:how-to/using-rocprofv3>`.

ROCprof Compute Viewer
----------------------

ROCprof Compute Viewer provides a GUI-based environment for analyzing GPU
kernel performance data. It specializes in kernel-level analysis, offering
counter correlation and hierarchical performance breakdowns that make it easier

> to understand kernel execution patterns and identify optimization opportunities.

Reviewer 3: ```suggestion
kernel performance data. It delivers detailed kernel‑level insights, including counter correlation and hierarchical performance breakdowns, helping developers interpret execution patterns and identify optimization opportunities.
```

Key features:

* Kernel performance counter analysis
* Counter correlation and visualization
* Hierarchical kernel performance breakdown
* Interactive performance data exploration
* Kernel-level optimization insights

For more information, see the
`ROCprof Compute Viewer documentation <https://rocm.docs.amd.com/projects/rocprof_compute_viewer/en/latest/>`__.

Perfetto trace viewer
---------------------

For application-level tracing, including API calls, library calls, and
host-device synchronization, use the Perfetto trace viewer at
https://ui.perfetto.dev. The ``rocprofv3`` tool can generate traces in
Perfetto-compatible format, allowing you to visualize the complete timeline of
application execution.

Perfetto provides a web-based interface for analyzing traces, showing the
temporal relationship between host operations, kernel launches, memory
transfers, and synchronization events. This holistic view helps identify
performance issues related to API overhead, inefficient synchronization, or

> suboptimal overlap between computation and data transfer.

Reviewer 3: ```suggestion
Perfetto provides a web‑based interface for analyzing traces, visualizing the temporal relationships among host operations, kernel launches, memory transfers, and synchronization events. This holistic perspective helps identify performance issues caused from API overhead, inefficient synchronization, or insufficient overlap between computation and data movement.
```
